File: dsa/examples/use-cases/code-signing/sign_release.py
#!/usr/bin/env python3
"""
Post-Quantum Code Signing Tool

Sign software releases (binaries, tarballs, packages) using post-quantum
digital signatures (ML-DSA or SLH-DSA) as specified in FIPS 204/205.

This tool creates detached signatures with metadata including:
- Timestamp (ISO 8601 format)
- Algorithm used
- Signer identity (from certificate)
- File hash (SHA-256)

Usage:
    python sign_release.py --key secret.key --file release.tar.gz [options]

Output:
    Creates release.tar.gz.sig (JSON signature file)
"""

import argparse
import hashlib
import json
import sys
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Optional

# Try to import from installed package first, fall back to source path
try:
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )
except ImportError:
    # Add parent directories to path for imports (development mode)
    sys.path.insert(0, str(Path(__file__).parent.parent.parent.parent / "src" / "py"))
    from mldsa import MLDSA44, MLDSA65, MLDSA87
    from slhdsa import (
        SLHDSA_SHAKE_128f, SLHDSA_SHAKE_128s,
        SLHDSA_SHAKE_192f, SLHDSA_SHAKE_256f,
    )


# Supported algorithms
ALGORITHMS = {
    "mldsa44": {"class": MLDSA44, "name": "ML-DSA-44", "security": "NIST Level 2"},
    "mldsa65": {"class": MLDSA65, "name": "ML-DSA-65", "security": "NIST Level 3"},
    "mldsa87": {"class": MLDSA87, "name": "ML-DSA-87", "security": "NIST Level 5"},
    "slh-shake-128f": {"class": SLHDSA_SHAKE_128f, "name": "SLH-DSA-SHAKE-128f", "security": "NIST Level 1"},
    "slh-shake-128s": {"class": SLHDSA_SHAKE_128s, "name": "SLH-DSA-SHAKE-128s", "security": "NIST Level 1"},
    "slh-shake-192f": {"class": SLHDSA_SHAKE_192f, "name": "SLH-DSA-SHAKE-192f", "security": "NIST Level 3"},
    "slh-shake-256f": {"class": SLHDSA_SHAKE_256f, "name": "SLH-DSA-SHAKE-256f", "security": "NIST Level 5"},
}


def compute_file_hash(filepath: Path, algorithm: str = "sha256") -> str:
    """Compute cryptographic hash of a file."""
    h = hashlib.new(algorithm)
    with open(filepath, "rb") as f:
        # Read in chunks for large files
        for chunk in iter(lambda: f.read(65536), b""):
            h.update(chunk)
    return h.hexdigest()


def load_secret_key(key_path: Path, password: Optional[str] = None) -> tuple[bytes, dict]:
    """
    Load secret key and certificate metadata.

    Returns:
        Tuple of (key_bytes, certificate_dict)
    """
    key_data = key_path.read_bytes()

    # Check for encrypted key (JSON format with encryption metadata)
    if key_data.startswith(b'{'):
        key_json = json.loads(key_data)
        if "encrypted" in key_json and key_json["encrypted"]:
            if not password:
                raise ValueError("Key is encrypted but no password provided")
            # Decrypt key (simplified - in production use proper KDF)
            raise NotImplementedError("Encrypted key support requires password decryption")
        key_data = bytes.fromhex(key_json.get("key", key_json.get("secret_key", "")))

    # Load certificate if exists
    cert_path = key_path.parent / key_path.name.replace("_secret.key", "_certificate.json")
    cert_data = {}
    if cert_path.exists():
        cert_data = json.loads(cert_path.read_text())

    return key_data, cert_data


def detect_algorithm(key_bytes: bytes) -> str:
    """Detect algorithm from secret key size."""
    key_sizes = {
        2560: "mldsa44",   # ML-DSA-44 secret key
        4032: "mldsa65",   # ML-DSA-65 secret key
        4896: "mldsa87",   # ML-DSA-87 secret key
    }

    size = len(key_bytes)
    if size in key_sizes:
        return key_sizes[size]

    # SLH-DSA keys are larger, check ranges
    if 64 <= size <= 128:
        # Could be SLH-DSA, need more context
        return "slh-shake-128f"  # Default guess

    raise ValueError(f"Cannot detect algorithm from key size {size} bytes")


def sign_file(
    filepath: Path,
    secret_key: bytes,
    algorithm: str,
    signer_info: dict,
    context: bytes = b"",
) -> dict:
    """
    Sign a file and return signature metadata.

    Args:
        filepath: Path to file to sign
        secret_key: Secret signing key bytes
        algorithm: Algorithm identifier (e.g., "mldsa65")
        signer_info: Dictionary with signer identity information
        context: Optional context string for domain separation

    Returns:
        Dictionary containing signature and metadata
    """
    if algorithm not in ALGORITHMS:
        raise ValueError(f"Unknown algorithm: {algorithm}")

    algo_info = ALGORITHMS[algorithm]

    # Compute file hash
    file_hash = compute_file_hash(filepath)
    file_size = filepath.stat().st_size

    # Generate timestamp once (used in both signed message and metadata)
    timestamp = datetime.now(timezone.utc).isoformat()

    # Create message to sign (hash + metadata for binding)
    # This prevents signature reuse across different files with same hash
    sign_message = json.dumps({
        "file_hash": file_hash,
        "file_name": filepath.name,
        "file_size": file_size,
        "timestamp": timestamp,
    }, sort_keys=True).encode("utf-8")

    # Initialize signer
    signer = algo_info["class"]()

    # Sign the message
    signature = signer.sign(secret_key, sign_message, ctx=context)

    # Build signature document
    sig_doc = {
        "version": "1.0",
        "type": "code-signature",
        "algorithm": {
            "id": algorithm,
            "name": algo_info["name"],
            "security_level": algo_info["security"],
            "standard": "FIPS 204" if algorithm.startswith("mldsa") else "FIPS 205",
        },
        "file": {
            "name": filepath.name,
            "size": file_size,
            "hash": {
                "algorithm": "sha256",
                "value": file_hash,
            },
        },
        "signature": {
            "value": signature.hex(),
            "encoding": "hex",
            "context": context.hex() if context else "",
        },
        "timestamp": timestamp,
        "signer": signer_info,
        "metadata": {
            "tool": "pqc-code-signing",
            "tool_version": "1.0.0",
        },
    }

    return sig_doc


def main():
    parser = argparse.ArgumentParser(
        description="Sign software releases with post-quantum signatures",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Sign a release tarball
  python sign_release.py --key mykey_secret.key --file release-1.0.tar.gz

  # Sign with specific algorithm
  python sign_release.py --key mykey_secret.key --file app.exe --algorithm mldsa87

  # Sign with context (domain separation)
  python sign_release.py --key mykey_secret.key --file firmware.bin --context "firmware-v2"

  # Specify output file
  python sign_release.py --key mykey_secret.key --file release.tar.gz --output release.tar.gz.pqsig
        """,
    )

    parser.add_argument(
        "--key", "-k",
        required=True,
        type=Path,
        help="Path to secret key file",
    )
    parser.add_argument(
        "--file", "-f",
        required=True,
        type=Path,
        help="Path to file to sign",
    )
    parser.add_argument(
        "--algorithm", "-a",
        choices=list(ALGORITHMS.keys()),
        help="Signing algorithm (auto-detected from key if not specified)",
    )
    parser.add_argument(
        "--output", "-o",
        type=Path,
        help="Output signature file (default: <file>.sig)",
    )
    parser.add_argument(
        "--context", "-c",
        default="",
        help="Context string for domain separation",
    )
    parser.add_argument(
        "--password", "-p",
        help="Password for encrypted key",
    )
    parser.add_argument(
        "--signer-name",
        help="Signer name (overrides certificate)",
    )
    parser.add_argument(
        "--signer-email",
        help="Signer email (overrides certificate)",
    )
    parser.add_argument(
        "--quiet", "-q",
        action="store_true",
        help="Suppress output except errors",
    )

    args = parser.parse_args()

    # Validate inputs
    if not args.key.exists():
        print(f"Error: Key file not found: {args.key}", file=sys.stderr)
        sys.exit(1)

    if not args.file.exists():
        print(f"Error: File not found: {args.file}", file=sys.stderr)
        sys.exit(1)

    # Load secret key
    try:
        secret_key, cert_data = load_secret_key(args.key, args.password)
    except Exception as e:
        print(f"Error loading key: {e}", file=sys.stderr)
        sys.exit(1)

    # Determine algorithm
    algorithm = args.algorithm
    if not algorithm:
        try:
            algorithm = detect_algorithm(secret_key)
            if not args.quiet:
                print(f"Auto-detected algorithm: {algorithm}")
        except ValueError as e:
            print(f"Error: {e}. Please specify --algorithm", file=sys.stderr)
            sys.exit(1)

    # Build signer info
    signer_info = {}
    if cert_data:
        subject = cert_data.get("subject", {})
        signer_info = {
            "common_name": args.signer_name or subject.get("common_name", ""),
            "organization": subject.get("organization", ""),
            "email": args.signer_email or subject.get("email", ""),
        }
    else:
        signer_info = {
            "common_name": args.signer_name or "Unknown",
            "email": args.signer_email or "",
        }

    # Remove empty fields
    signer_info = {k: v for k, v in signer_info.items() if v}

    # Sign the file
    try:
        sig_doc = sign_file(
            filepath=args.file,
            secret_key=secret_key,
            algorithm=algorithm,
            signer_info=signer_info,
            context=args.context.encode("utf-8"),
        )
    except Exception as e:
        print(f"Error signing file: {e}", file=sys.stderr)
        sys.exit(1)

    # Write signature file
    output_path = args.output or args.file.with_suffix(args.file.suffix + ".sig")
    try:
        with open(output_path, "w") as f:
            json.dump(sig_doc, f, indent=2)
    except Exception as e:
        print(f"Error writing signature: {e}", file=sys.stderr)
        sys.exit(1)

    if not args.quiet:
        print(f"\nSignature created successfully!")
        print(f"  File:      {args.file}")
        print(f"  Hash:      {sig_doc['file']['hash']['value'][:16]}...")
        print(f"  Algorithm: {sig_doc['algorithm']['name']}")
        print(f"  Security:  {sig_doc['algorithm']['security_level']}")
        print(f"  Output:    {output_path}")


if __name__ == "__main__":
    main()

File: dsa/examples/use-cases/code-signing/sign_release.cpp
/**
 * Post-Quantum Code Signing Tool (C++)
 *
 * Signs software releases using ML-DSA or SLH-DSA, producing a detached
 * JSON signature manifest for distribution.
 *
 * Usage:
 *   ./sign_release <secret_key> <file> [options]
 *
 * Options:
 *   --algorithm <alg>      Algorithm (auto-detected from key size if omitted)
 *   --output <file>        Output signature file (default: <file>.sig)
 *   --context <str>        Context string for domain separation
 *   --signer-name <name>   Signer name
 *   --signer-email <email> Signer email
 *   --quiet                Suppress output
 *
 * Example:
 *   ./sign_release keys/release_secret.key dist/myapp-1.0.0.tar.gz
 */

#include <iostream>
#include <fstream>
#include <sstream>
#include <iomanip>
#include <chrono>
#include <vector>
#include <string>
#include <algorithm>
#include <stdexcept>
#include <cstring>

#include "mldsa/mldsa.hpp"
#include "slhdsa/slh_dsa.hpp"

// SHA-256 implementation (simplified for demonstration)
#include <openssl/sha.h>
#include <openssl/evp.h>

// Read binary file
std::vector<uint8_t> read_file(const std::string& path) {
    std::ifstream file(path, std::ios::binary);
    if (!file) {
        throw std::runtime_error("Cannot open file: " + path);
    }
    return std::vector<uint8_t>(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>()
    );
}

// Write text file
bool write_text_file(const std::string& path, const std::string& content) {
    std::ofstream file(path);
    if (!file) return false;
    file << content;
    return file.good();
}

// Convert bytes to hex string
std::string to_hex(const std::vector<uint8_t>& data) {
    std::ostringstream oss;
    for (uint8_t b : data) {
        oss << std::hex << std::setfill('0') << std::setw(2) << static_cast<int>(b);
    }
    return oss.str();
}

// Compute SHA-256 hash
std::vector<uint8_t> sha256(const std::vector<uint8_t>& data) {
    std::vector<uint8_t> hash(SHA256_DIGEST_LENGTH);
    SHA256(data.data(), data.size(), hash.data());
    return hash;
}

// Get current ISO timestamp
std::string get_timestamp() {
    auto now = std::chrono::system_clock::now();
    auto time = std::chrono::system_clock::to_time_t(now);
    std::stringstream ss;
    ss << std::put_time(std::gmtime(&time), "%Y-%m-%dT%H:%M:%S+00:00");
    return ss.str();
}

// Extract filename from path
std::string get_filename(const std::string& path) {
    size_t pos = path.find_last_of("/\\");
    return (pos == std::string::npos) ? path : path.substr(pos + 1);
}

// JSON escape string
std::string json_escape(const std::string& s) {
    std::string result;
    for (char c : s) {
        switch (c) {
            case '"': result += "\\\""; break;
            case '\\': result += "\\\\"; break;
            case '\n': result += "\\n"; break;
            case '\r': result += "\\r"; break;
            case '\t': result += "\\t"; break;
            default: result += c;
        }
    }
    return result;
}

// Algorithm info structure
struct AlgorithmInfo {
    std::string id;
    std::string name;
    std::string security_level;
    std::string standard;
    size_t pk_size;
    size_t sk_size;
};

// Get algorithm info from name
AlgorithmInfo get_algorithm_info(const std::string& alg) {
    std::string lower = alg;
    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);

    if (lower == "mldsa44") {
        return {"mldsa44", "ML-DSA-44", "NIST Level 2", "FIPS 204", 1312, 2560};
    } else if (lower == "mldsa65") {
        return {"mldsa65", "ML-DSA-65", "NIST Level 3", "FIPS 204", 1952, 4032};
    } else if (lower == "mldsa87") {
        return {"mldsa87", "ML-DSA-87", "NIST Level 5", "FIPS 204", 2592, 4896};
    } else if (lower == "slh-shake-128f") {
        return {"slh-shake-128f", "SLH-DSA-SHAKE-128f", "NIST Level 1", "FIPS 205", 32, 64};
    } else if (lower == "slh-shake-192f") {
        return {"slh-shake-192f", "SLH-DSA-SHAKE-192f", "NIST Level 3", "FIPS 205", 48, 96};
    } else if (lower == "slh-shake-256f") {
        return {"slh-shake-256f", "SLH-DSA-SHAKE-256f", "NIST Level 5", "FIPS 205", 64, 128};
    }
    throw std::invalid_argument("Unknown algorithm: " + alg);
}

// Detect algorithm from secret key size
std::string detect_algorithm(size_t sk_size) {
    switch (sk_size) {
        case 2560: return "mldsa44";
        case 4032: return "mldsa65";
        case 4896: return "mldsa87";
        case 64: return "slh-shake-128f";
        case 96: return "slh-shake-192f";
        case 128: return "slh-shake-256f";
        default:
            throw std::invalid_argument("Cannot detect algorithm from key size: " +
                                        std::to_string(sk_size));
    }
}

// Sign function template for ML-DSA
template<typename DSA>
std::vector<uint8_t> sign_mldsa(const std::vector<uint8_t>& sk,
                                 const std::vector<uint8_t>& message,
                                 const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.sign(sk, message, ctx);
}

// Sign function template for SLH-DSA
template<typename DSA>
std::vector<uint8_t> sign_slhdsa(const std::vector<uint8_t>& sk,
                                  const std::vector<uint8_t>& message,
                                  const std::vector<uint8_t>& ctx) {
    DSA dsa;
    return dsa.sign(sk, message, ctx);
}

// Create signature manifest JSON
std::string create_manifest(const std::string& filename,
                           size_t file_size,
                           const std::string& file_hash,
                           const std::string& timestamp,
                           const AlgorithmInfo& alg,
                           const std::vector<uint8_t>& signature,
                           const std::string& context,
                           const std::string& signer_name,
                           const std::string& signer_email) {
    std::ostringstream json;
    json << "{\n";
    json << "  \"manifest_version\": \"1.0\",\n";
    json << "  \"type\": \"code-signature\",\n";
    json << "  \"algorithm\": {\n";
    json << "    \"id\": \"" << alg.id << "\",\n";
    json << "    \"name\": \"" << alg.name << "\",\n";
    json << "    \"security_level\": \"" << alg.security_level << "\",\n";
    json << "    \"standard\": \"" << alg.standard << "\"\n";
    json << "  },\n";
    json << "  \"file\": {\n";
    json << "    \"name\": \"" << json_escape(filename) << "\",\n";
    json << "    \"size\": " << file_size << ",\n";
    json << "    \"hash\": \"" << file_hash << "\",\n";
    json << "    \"hash_algorithm\": \"sha256\"\n";
    json << "  },\n";
    json << "  \"signature\": {\n";
    json << "    \"value\": \"" << to_hex(signature) << "\",\n";
    json << "    \"encoding\": \"hex\"";
    if (!context.empty()) {
        json << ",\n    \"context\": \"" << to_hex(std::vector<uint8_t>(context.begin(), context.end())) << "\"";
    }
    json << "\n  },\n";
    json << "  \"timestamp\": \"" << timestamp << "\"";

    if (!signer_name.empty() || !signer_email.empty()) {
        json << ",\n  \"signer\": {\n";
        if (!signer_name.empty()) {
            json << "    \"name\": \"" << json_escape(signer_name) << "\"";
            if (!signer_email.empty()) json << ",\n";
            else json << "\n";
        }
        if (!signer_email.empty()) {
            json << "    \"email\": \"" << json_escape(signer_email) << "\"\n";
        }
        json << "  }";
    }

    json << "\n}\n";
    return json.str();
}

void print_usage() {
    std::cout << "Post-Quantum Code Signing Tool (C++)\n";
    std::cout << std::string(60, '=') << "\n\n";
    std::cout << "Usage: sign_release <secret_key> <file> [options]\n\n";
    std::cout << "Required:\n";
    std::cout << "  <secret_key>          Secret key file\n";
    std::cout << "  <file>                File to sign\n\n";
    std::cout << "Options:\n";
    std::cout << "  --algorithm <alg>     Algorithm (auto-detected from key)\n";
    std::cout << "  --output <file>       Output signature file (default: <file>.sig)\n";
    std::cout << "  --context <str>       Context string for domain separation\n";
    std::cout << "  --signer-name <name>  Signer name\n";
    std::cout << "  --signer-email <email> Signer email\n";
    std::cout << "  --quiet               Suppress output\n";
    std::cout << "  --help                Show this help\n\n";
    std::cout << "Algorithms:\n";
    std::cout << "  mldsa44, mldsa65, mldsa87 (ML-DSA, FIPS 204)\n";
    std::cout << "  slh-shake-128f, slh-shake-192f, slh-shake-256f (SLH-DSA, FIPS 205)\n\n";
    std::cout << "Example:\n";
    std::cout << "  ./sign_release keys/release_secret.key dist/myapp-1.0.0.tar.gz\n";
}

int main(int argc, char* argv[]) {
    if (argc < 3) {
        print_usage();
        return (argc == 1) ? 0 : 1;
    }

    std::string sk_file;
    std::string file_path;
    std::string algorithm;
    std::string output_file;
    std::string context;
    std::string signer_name;
    std::string signer_email;
    bool quiet = false;

    // Parse arguments
    int i = 1;
    while (i < argc) {
        std::string arg = argv[i];

        if (arg == "--algorithm" && i + 1 < argc) {
            algorithm = argv[++i];
        } else if (arg == "--output" && i + 1 < argc) {
            output_file = argv[++i];
        } else if (arg == "--context" && i + 1 < argc) {
            context = argv[++i];
        } else if (arg == "--signer-name" && i + 1 < argc) {
            signer_name = argv[++i];
        } else if (arg == "--signer-email" && i + 1 < argc) {
            signer_email = argv[++i];
        } else if (arg == "--quiet" || arg == "-q") {
            quiet = true;
        } else if (arg == "--help" || arg == "-h") {
            print_usage();
            return 0;
        } else if (arg[0] != '-') {
            if (sk_file.empty()) {
                sk_file = arg;
            } else if (file_path.empty()) {
                file_path = arg;
            }
        } else {
            std::cerr << "Unknown option: " << arg << "\n";
            return 1;
        }
        ++i;
    }

    if (sk_file.empty() || file_path.empty()) {
        std::cerr << "Error: Secret key and file are required\n";
        return 1;
    }

    try {
        // Read secret key
        if (!quiet) std::cerr << "Reading secret key: " << sk_file << "\n";
        auto sk = read_file(sk_file);

        // Auto-detect algorithm if not specified
        if (algorithm.empty()) {
            algorithm = detect_algorithm(sk.size());
            if (!quiet) std::cerr << "Detected algorithm: " << algorithm << "\n";
        }

        // Read file to sign
        if (!quiet) std::cerr << "Reading file: " << file_path << "\n";
        auto file_data = read_file(file_path);

        // Compute hash
        auto hash = sha256(file_data);
        std::string hash_hex = to_hex(hash);

        // Get timestamp
        std::string timestamp = get_timestamp();

        // Build signed message (hash + filename + size + timestamp)
        std::string filename = get_filename(file_path);
        std::ostringstream msg_stream;
        msg_stream << "{\"file_hash\":\"" << hash_hex << "\",";
        msg_stream << "\"file_name\":\"" << json_escape(filename) << "\",";
        msg_stream << "\"file_size\":" << file_data.size() << ",";
        msg_stream << "\"timestamp\":\"" << timestamp << "\"}";
        std::string msg_str = msg_stream.str();
        std::vector<uint8_t> message(msg_str.begin(), msg_str.end());

        // Prepare context
        std::vector<uint8_t> ctx_bytes(context.begin(), context.end());

        // Sign
        if (!quiet) std::cerr << "Signing with " << algorithm << "...\n";
        std::vector<uint8_t> signature;
        AlgorithmInfo alg_info = get_algorithm_info(algorithm);

        if (algorithm == "mldsa44") {
            signature = sign_mldsa<mldsa::MLDSA44>(sk, message, ctx_bytes);
        } else if (algorithm == "mldsa65") {
            signature = sign_mldsa<mldsa::MLDSA65>(sk, message, ctx_bytes);
        } else if (algorithm == "mldsa87") {
            signature = sign_mldsa<mldsa::MLDSA87>(sk, message, ctx_bytes);
        } else if (algorithm == "slh-shake-128f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_128f>(sk, message, ctx_bytes);
        } else if (algorithm == "slh-shake-192f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_192f>(sk, message, ctx_bytes);
        } else if (algorithm == "slh-shake-256f") {
            signature = sign_slhdsa<slhdsa::SLHDSA_SHAKE_256f>(sk, message, ctx_bytes);
        } else {
            throw std::invalid_argument("Unsupported algorithm: " + algorithm);
        }

        if (!quiet) std::cerr << "Signature size: " << signature.size() << " bytes\n";

        // Create manifest
        std::string manifest = create_manifest(
            filename, file_data.size(), hash_hex, timestamp,
            alg_info, signature, context, signer_name, signer_email
        );

        // Write output
        if (output_file.empty()) {
            output_file = file_path + ".sig";
        }

        if (!write_text_file(output_file, manifest)) {
            std::cerr << "Error: Failed to write signature file\n";
            return 1;
        }

        if (!quiet) {
            std::cerr << "Signature written to: " << output_file << "\n";
            std::cerr << "Signing complete.\n";
        }

        return 0;

    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << "\n";
        return 1;
    }
}
